### DISCUSSION TITLE
Will NPM force private packages to be scoped at some point ?

### DISCUSSION THREAD

**Discussion by rvillane** [npm/feedback#880] [General]
The NPM roadmap [1] is not showing this level of detail.  My understanding is that using NPM scoped packages is a best practice and now NPM 9 enforces scoped auth credentials and seems like eventually, scopes might be forced but cannot find any reference that confirms if this statement is accurate.

Any thoughts ?

[1] https://github.com/orgs/github/projects/4247/views/1?filterQuery=npm

**Comment by ljharb** [comment]
Why would scoped ever be forced? Unscoped package names aren’t a bad practice.

**Reply by rvillane** [reply]
that is what I'm trying to figure out :).  I don't know if is a bad practice but it opens the door to NPM substitution attacks [1] and scoped packages is one way of remediating.

[1] https://github.blog/2021-02-12-avoiding-npm-substitution-attacks/

**Reply by ljharb** [reply]
That door is only open if you're failing to follow years of best practices - certainly all your *internally published* packages should be scoped, but *private* packages are ones that aren't published at all, and there's not really any reliable programmatic way for npm to know which packages are internally vs publicly published.

**Comment by rvillane** [comment]
that is an interesting distinction internal vs private, that concept is new to me, in my short understanding we have:

1. public NPM packages, those published on a public NPM registry
2. private NPM packages, those published on a private NPM registry

I'm not really familiar with internal packages, and the intention of this discussion is focused around 2) Npm libraries published on private registries

**Reply by ljharb** [reply]
right. since package.json has a `private` field, for things that aren't published at all, the nomenclature is a bit muddled.

I'm calling number 2 there "internal packages" to disambiguate from `private: true` packages.

**Comment by rvillane** [comment]
Now that we are on the same page regarding terminology and going back to the original question, is NPM heading to force scopes on internal packages at some point in the future?

**Reply by ljharb** [reply]
In order to do that, npm would have to have a way to programmatically know when a package is internal. How do you suggest detecting that?

**Reply by rvillane** [reply]
just by the fact is getting published to a private repo perhaps

**Reply by ljharb** [reply]
how can npm know that tho? sometimes the destination registry is provided on the CLI with the publish command, or maybe via env var in CI (and never on a developer's machine). Sometimes an alternate registry will automatically republish to the public registry. Sometimes a package is published *twice* - once internally and once publicly.

**Reply by rvillane** [reply]
very interesting, seems like scopes is just a recommended practice, which is what I was looking to understand.

thanks for the chat, very enlightening
